storage/innobase/row/row0ins.cc

```diff
@@ -2703,22 +2703,43 @@ row_ins_clust_index_entry_low(
 
 	DBUG_EXECUTE_IF("row_ins_row_level", goto skip_bulk_insert;);
 
-	if (!(flags & BTR_NO_UNDO_LOG_FLAG)
-	    && page_is_empty(block->page.frame)
-	    && !entry->is_metadata() && !trx->duplicates
-	    && !trx->check_unique_secondary && !trx->check_foreigns
-	    && !trx->dict_operation
-	    && block->page.id().page_no() == index->page
-	    && !index->table->skip_alter_undo
+	if (((!(flags & BTR_NO_UNDO_LOG_FLAG)
+	      && page_is_empty(block->page.frame)
+	      && !entry->is_metadata() && !trx->duplicates
+	      && !trx->check_unique_secondary && !trx->check_foreigns
+	      && !trx->dict_operation
+	      && !index->table->is_active_ddl()
+	      && block->page.id().page_no() == index->page)
+	     || (index->table->skip_alter_undo
+		 && !index->table->is_temporary()
+		 && index->table->foreign_set.empty()))
```

Comment:
As far as I can tell, `flags == BTR_NO_UNDO_LOG_FLAG | BTR_NO_LOCKING_FLAG` in the case that we are interested in. That is, the condition `index->table->skip_alter_undo` could be replaced with that. But, we can do even better.

If we introduced one more `flags` bit and set it in `row_ins_clust_index_entry()` appropriately, we could replace all these 3 conditions with just one `flags ==` test.